Bug 1876328: Update virtualization os images namespace

yaacov opened the pull request:
Base images namespace name for CNV was updated from
"openshift-kubevirt-base-images" to "openshift-virtualization-os-images"

This PR updates the namespace in the UI to use the template namesapce instead of using a hardcoded namespace

yaacov commented:
@rawagner @glekner @omeryahud please review

omeryahud commented on frontend/packages/kubevirt-plugin/src/constants/vm/constants.ts:
```diff
@@ -25,7 +25,7 @@ export const TEMPLATE_DATAVOLUME_NAME_PARAMETER = 'SRC_PVC_NAME';
 export const TEMPLATE_VM_DOMAIN_LABEL = 'kubevirt.io/domain';
 export const TEMPLATE_VM_SIZE_LABEL = 'kubevirt.io/size';
-export const TEMPLATE_VM_GOLDEN_OS_NAMESPACE = 'openshift-kubevirt-base-images';
+export const TEMPLATE_VM_GOLDEN_OS_NAMESPACE = 'openshift-virtualization-os-images';
```
Aren't you supposed to fetch this namespace name from our templates instead of declaring it yourselves?

yaacov replied:
In the UI we list all the OSs with their os PVC, so we need to loop over each OS, parse the name, namespace for the PVC and check against a list of PVCs, assuming all the os-images are in one namespace saves us the trouble of going over all the PVCs in the cluster or queering for individual PVC for each OS.

@glekner @rawagner - 
should we switch to fetching all PVCs in all nameapces ?
do we have a way, while in a component, to query for individual PVC per OS so we will not have to get a list of all PVCs and search for it in our code ?

rawagner replied:
we can do a request for specific PVC with given name/namespace once user picks OS.

yaacov replied:
>we can do a request for specific PVC with given name/namespace once user picks OS.

we do that, the problematic places are the dropdowns we show before selecting an OS in this dropdowns we show a list of all OS with a note saying if we found a matching PVC or not, to solve this we need to loop over each OS and query if it has a matching PVC, It's hard to do using hooks and/or firehose ... @rawagner can you think of a way to implement this dropdowns in a nicer way ?

yaacov commented:
Changing to WIP because we will use this PR to not requiring a hardcoded namespace
cc// @rawagner @glekner @omeryahud

yaacov commented:
WIP again :-) we should not use all-namespaces

yaacov commented:
a. added a hook to get the PVC's that are used by common templates.
b. converted the vm create wizard from object to function to use the hook.

@rawagner @glekner please review.

rawagner commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/create-vm-wizard.tsx:
```diff
@@ -78,217 +80,230 @@ type CreateVMWizardComponentProps = {
+      if (dataPVCsChanged) {
+        commonDataUpdate = {
+          ...commonDataUpdate,
+          data: {
```
```suggestion
          data: {
            ...commonDataUpdate.data,
```

yaacov replied:
he :-) fixed

yaacov commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/create-vm-wizard.tsx:
```diff
@@ -78,217 +80,230 @@ type CreateVMWizardComponentProps = {
-  private isClosed = false;
+const CreateVMWizardComponent: React.FC<CreateVMWizardComponentProps> = (props) => {
+  const [closed, setClosed] = React.useState(false);
+  const [dataVolumePVCs, dataVolumePVCsLoaded, dataVolumePVCsLoadError] = usePVCBaseImages(
```
thanks !

yaacov commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/types.ts:
```diff
@@ -269,6 +269,11 @@ export type CommonData = {
+    [VMWizardProps.openshiftCNVBaseImages]?: {
+      loaded: boolean;
+      loadError: string;
+      data: any[];
```
:-)

yaacov commented on frontend/packages/kubevirt-plugin/src/hooks/use-pvc-base-images.ts:
```diff
@@ -0,0 +1,50 @@
+import { getLoadedData, getLoadError, isLoaded } from '../utils';
+import { PersistentVolumeClaimModel } from '@console/internal/models';
+
+export const usePVCBaseImages = (commonTemplates: any[]) => {
```
wow, nice !

yaacov commented on frontend/packages/kubevirt-plugin/src/components/cdi-upload-provider/upload-pvc-form/upload-pvc-form.tsx:
```diff
@@ -544,7 +547,7 @@ export type UploadPVCFormProps = {
   isLoading: boolean;
   commonTemplates: TemplateKind[];
-  goldenPvcs: K8sResourceKind[];
+  goldenPvcs: any;
```
sorry :-) fixed

yaacov commented:
@rawagner hi, thanks for the review :+1: I think I fixed all the things, please re-review

rawagner commented:
/lgtm
